File: softprobe-js/src/__tests__/parity-dogfood.test.ts
/**
 * PD7.3a — TS SDK parity test.
 *
 * Drives the full Softprobe facade (startSession → loadCaseFromFile →
 * findInCase → mockOutbound → close) against a fake runtime using the
 * checked-in golden case `fragment-happy-path.case.json`.
 *
 * This test proves that the TS SDK correctly serialises every control-API
 * call without a live runtime or proxy.
 */
import path from 'path';

import { Softprobe } from '../softprobe';

const GOLDEN_CASE = path.join(
  __dirname,
  '..', '..', '..', 'spec', 'examples', 'cases', 'fragment-happy-path.case.json',
);

describe('TS SDK parity (PD7.3a)', () => {
  it('drives startSession → loadCaseFromFile → findInCase → mockOutbound → close', async () => {
    const calls: Array<{ method: string; url: string; body: unknown }> = [];

    const fetchImpl = async (input: string, init?: { method?: string; body?: string }) => {
      const url = String(input);
      const method = (init?.method ?? 'GET').toUpperCase();
      const body = init?.body ? JSON.parse(init.body) : {};
      calls.push({ method, url, body });

      if (url.endsWith('/v1/sessions') && method === 'POST') {
        return {
          ok: true,
          status: 200,
          text: async () => JSON.stringify({ sessionId: 'dogfood-session', sessionRevision: 0 }),
        };
      }
      if (url.includes('/load-case') && method === 'POST') {
        return {
          ok: true,
          status: 200,
          text: async () => JSON.stringify({ sessionId: 'dogfood-session', sessionRevision: 1 }),
        };
      }
      if (url.includes('/rules') && method === 'POST') {
        return {
          ok: true,
          status: 200,
          text: async () => JSON.stringify({ sessionId: 'dogfood-session', sessionRevision: 2 }),
        };
      }
      if (url.endsWith('/close') && method === 'POST') {
        return {
          ok: true,
          status: 200,
          text: async () => JSON.stringify({ sessionId: 'dogfood-session', closed: true }),
        };
      }
      return { ok: false, status: 500, text: async () => `unexpected: ${method} ${url}` };
    };

    const sp = new Softprobe({ baseUrl: 'http://fake-runtime.test', fetchImpl });

    // start
    const session = await sp.startSession({ mode: 'replay' });
    expect(session.id).toBe('dogfood-session');

    // loadCaseFromFile — reads the actual golden JSON from disk
    await session.loadCaseFromFile(GOLDEN_CASE);

    // findInCase — in-memory lookup; no HTTP call
    const hit = session.findInCase({
      direction: 'outbound',
      method: 'GET',
      path: '/fragment',
    });
    expect(hit).toBeDefined();
    expect(hit.response).toBeDefined();

    // mockOutbound — posts the rule to the fake runtime
    await session.mockOutbound({
      direction: 'outbound',
      method: 'GET',
      path: '/fragment',
      response: hit.response,
    });

    // close
    await session.close();

    // Verify the sequence of calls against the fake runtime.
    const urls = calls.map((c) => c.url);
    expect(urls).toContain('http://fake-runtime.test/v1/sessions');
    expect(urls).toContain('http://fake-runtime.test/v1/sessions/dogfood-session/load-case');
    expect(urls.some((u) => u.includes('/rules'))).toBe(true);
    expect(urls.some((u) => u.endsWith('/close'))).toBe(true);

    // mockOutbound rule must reference /fragment
    const rulesCall = calls.find((c) => c.url.includes('/rules') && c.method === 'POST');
    expect(rulesCall).toBeDefined();
    const rulesBody = rulesCall!.body as { version: number; rules: unknown[] };
    expect(rulesBody.rules.length).toBeGreaterThanOrEqual(1);
  });
});

File: softprobe-java/src/test/java/com/softprobe/ParityDogfoodTest.java
package com.softprobe;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.net.URI;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

/**
 * PD7.3c — Java SDK parity test.
 *
 * Drives the full Softprobe facade (startSession → loadCaseFromFile →
 * findInCase → mockOutbound → close) against a fake transport using the
 * checked-in golden case fragment-happy-path.case.json.
 */
class ParityDogfoodTest {

  private static final Path GOLDEN_CASE =
      Paths.get("../spec/examples/cases/fragment-happy-path.case.json").toAbsolutePath();

  private static final class RecordingTransport implements Client.Transport {
    final List<Client.Request> calls = new ArrayList<>();
    int revision = 0;

    @Override
    public Client.Response send(Client.Request request) {
      calls.add(request);
      if (request.uri().getPath().endsWith("/close")) {
        return new Client.Response(200, "{\"sessionId\":\"dogfood-session\",\"closed\":true}");
      }
      revision += 1;
      return new Client.Response(
          200,
          "{\"sessionId\":\"dogfood-session\",\"sessionRevision\":" + revision + "}");
    }
  }

  @Test
  void fullFacadeAgainstGoldenCase() throws Exception {
    RecordingTransport transport = new RecordingTransport();
    Softprobe softprobe = new Softprobe("http://fake-runtime.test", transport);

    SoftprobeSession session = softprobe.startSession("replay");
    assertEquals("dogfood-session", session.id());

    session.loadCaseFromFile(GOLDEN_CASE);

    CapturedHit hit =
        session.findInCase(
            new CaseSpanPredicate().direction("outbound").method("GET").path("/fragment"));
    assertNotNull(hit);
    assertNotNull(hit.response());

    session.mockOutbound(
        new MockRuleSpec()
            .name("fragment-replay")
            .direction("outbound")
            .method("GET")
            .path("/fragment")
            .response(hit.response()));

    session.close();

    // Verify call sequence: create, load-case, rules, close
    List<URI> uris = new ArrayList<>();
    for (Client.Request call : transport.calls) {
      uris.add(call.uri());
    }
    assertTrue(uris.stream().anyMatch(u -> u.getPath().endsWith("/v1/sessions")));
    assertTrue(uris.stream().anyMatch(u -> u.getPath().endsWith("/load-case")));
    assertTrue(uris.stream().anyMatch(u -> u.getPath().endsWith("/rules")));
    assertTrue(uris.stream().anyMatch(u -> u.getPath().endsWith("/close")));

    // rules call must carry the /fragment rule
    Client.Request rulesCall =
        transport.calls.stream()
            .filter(c -> c.uri().getPath().endsWith("/rules"))
            .findFirst()
            .orElseThrow();
    assertTrue(rulesCall.body().contains("/fragment"));
  }
}
